Lib/ntpath.py

```diff
@@ -549,7 +557,12 @@ def _readlink_deep(path, seen=None):
         while normcase(path) not in seen:
             seen.add(normcase(path))
             try:
+                old_path = path
                 path = _nt_readlink(path)
+                # Links may be relative, so resolve them against their
+                # own location
+                if not isabs(path):
+                    path = normpath(join(dirname(old_path), path))
```

Comment:
I pointed this problem out in a comment on the original pull request. The solution is not quite this simple because of ".." components in relative symlinks (e.g. "bar\\..\\spam"). The kernel resolves an absolute directory symbolic links (but not a junction mount point) to its target before it evaluates a ".." component in a relative link. For example, if "bar" -> "M:\\eggs\\foo", then  "bar\\..\\spam" resolves to "M:\\eggs\\spam". Similarly if `old_path` contains directory symlinks, the kernel will resolve them before evaluating the relative symlink. Also, unlike Unix, just leaving the ".." components in place doesn't help in Windows. The internal `CreateFileW` call in `nt.readlink` implements the same naive behavior when it resolves ".." components to create a native NT path. (NT doesn't generally support ".." components in file-system paths. Only the I/O manager's internal handling of `IO_REPARSE_TAG_SYMLINK` reparse points supports them.) 

If we're doing this manually, then we have to carefully copy how the kernel behaves. But I'm not sure how to go about that without a complete redesign to basically copy the POSIX `realpath`, with added complexity for the complications of Windows paths.

Reply:
Also, in the previous GitHub comment on the original PR, I pointed out that it's wrong to read and evaluate a remote mount point (junction) on the client. They can only be evaluated on the server. If there's a junction to volume "M:" in a remote path, that's the remote system's "M:" volume, not the client's "M:", if that even exists. If it's a symlink, on the other hand, then it's always evaluated on the client, and in that case it must be the client's "M:" volume. But even that has a complication. The redirector on the client dynamically adheres to the client's L2L, L2R, R2L, and R2R symlink policies, as defined in the registry at the time of the request. If remote symlinks are disallowed by R2L (remote-to-local, i.e. a remote link that targets a local path) or R2R (remote-to-remote, i.e. a remote link that targets a remote path), then opening the path fails with access denied. 

If we're manually following links, then we need to query these policy values (not difficult) and apply them. If we know that the root path is already resolved from subst/mapped drives, then we can just check for a UNC path to determine whether or not a symlink or junction is remote. Then always stop if it's a remote junction, and conditionally stop on a symlink if it's denied by one of the four symlink policies.

Reply:
So given the choice between making this "good enough" (considering it's been broken in most of these cases forever and nobody has complained) and rewriting it all, I'm going to go with good enough.

I'm pretty sure this handles the `bar\..\spam` case you mention though:

```
normpath(join(dirname("M:\eggs\foo"), "bar\..\spam"))
normpath(join("M:\eggs", "bar\..\spam"))
normpath("M:\eggs\bar\..\spam")
"M:\eggs\spam"
```

Or were you illustrating what _this_ code does and saying it's wrong? Is it meant to resolve to `spam` rather than ever looking at `bar` (i.e. call normpath first)?

Reply:
I think the way I initially approached `realpath` was adequate, and it didn't have these complications. I basically expanded on your implementation of `resolve` from pathlib. That implementation just walks the path in reverse until `_getfinalpathname` succeeds and then appends the unresolved components. This lets the kernel (I/O manager, filesystem filters, and SMB redirector and remote server) handle all reparsing and filesystem policies. 

The only advantage that I see for manually implementing `realpath` with our own path parsing would be to get a real path that reflects how the kernel tracks ancestor mount points (junctions) in order to handle them as regular directories when resolving a relative symlink that has ".." components. But that requires too much code and depends too much on poorly documented behavior.

> Or were you illustrating what this code does and saying it's wrong? Is it meant to resolve to spam rather than ever looking at bar (i.e. call normpath first)?

If either `old_path` or `path` contains a symlink to another directory (relative or absolute), then the I/O manager in the kernel resolves all symlinks in the path that come before a ".." component is evaluated. Thus we cannot evaluate either `normpath(dirname(old_path))` or `normpath(path)` and get anything remotely like a correct result. Similarly, leaving the ".." components in place will be just as wrong (in sharp contrast to how we normally address this problem in Unix) because `CreateFileW` works just as naively as `normpath` in this case. The only way either of those is valid is if all symlinks and subst/mapped drives in `old_path` and `path` are already resolved. 

Mount points (junctions) are a particularly hard complication in this case because, while they're resolved to the target as the kernel parses a path, they're also progressively tracked to make them behave like regular directories  in relative symlinks. For example, if "relative_symlink" is "..\\bar", then "C:\\foo\\junction\\relative_symlink" will resolve in the kernel to "C:\\foo\\bar" -- regardless of whatever "junction" resolves to. On the other hand, "C:\\foo\\absolute_symlink\\relative_symlink" resolves according to the target of "absolute_symlink". For example, if "absolute_symlink" -> "M:\\spam\\eggs", then the resolved path will be "M:\\spam\\bar". 

Plus there's the added complexity of identifying remote symlinks/junctions and remote symlink targets. We cannot follow a junction in a remote path (e.g. "\\server\share\spam\junction"). And following any symlink depends on policies set in the registry for local-to-local, local-to-remote, remote-to-local, and remote-to-remote behavior. By default, following any remote symlink is forbidden, whether the target is a local path (i.e. R2L) or a UNC path (i.e. R2R), but system policy can even forbid evaluating local symlinks that target local paths (L2L) and UNC paths (L2R).

Reply:
> On the other hand, "C:\foo\absolute_symlink\relative_symlink" resolves according to the target of "absolute_symlink". For example, if "absolute_symlink" -> "M:\spam\eggs", then the resolved path will be "M:\spam\bar".

Just to make sure I'm understanding, this assumes that `M:\spam\eggs\relative_symlink` also exists and is the same as the other `relative_symlink`, right?

This approach still walks backwards relying on `_getfinalpathname` - the primary difference being that it will also `readlink` to follow "real-enough" paths that can't be resolved by the IO manager. So I think this is good enough for the case where `_getfinalpathname` fails.

Reply:
Thinking more about the failure case, the most likely one is a dangling relative symlink that goes up multiple levels past another symlink, which would result in the relative symlink being joined to the parent directory without resolving the parent's one first.

Given how hard that scenario would be to construct, and the fact that it's purely theoretical and AFAIK nobody has ever hit it, I'm prepared to declare it a known limitation.

Reply:
> > On the other hand, "C:\foo\absolute_symlink\relative_symlink" resolves according to the target of "absolute_symlink". For example, if "absolute_symlink" -> "M:\spam\eggs", then the resolved path will be "M:\spam\bar".
> 
> Just to make sure I'm understanding, this assumes that `M:\spam\eggs\relative_symlink` also exists and is the same as the other `relative_symlink`, right?

Yes, it was meant to be the same "..\\bar" target. The final path is completely different depending on whether we reach "relative_symlink" by way of a junction or a symlink. But don't let that distract you from the point here about using `normpath`. The difference between junctions and symlinks is vital when manually parsing paths, but the primary issue I'm discussing is even more basic. 

In user mode, ".." components are handled naively. If we open "C:\\foo\\link\\..\\bar", the user-mode runtime library simply resolves this as "C:\\foo\\bar", just as `normpath` does. It doesn't care or know whether "link" is a junction, symlink, or any other type of reparse point. On the other hand, when reparsing a relative symlink in the kernel, these ".." components are processed in sequence, as the path is parsed. So "link" will already be evaluated to its target by the time the "..\\bar" relative symlink is parsed. Additionally, the complication that I mentioned is that if "link" is a mount point (junction), the system goes out of its way to remember the original path when grafting a relative symlink for a reparse. 

(Bear in mind that it has to resolve "link" to its target path in order to reach the relative symlink. So it's carrying forward the history of all junctions that parsing has traversed in order to logically implement ancestor ".." traversal for mount points. This is complicated, and I see several functions in the kernel symbol table that implement this behavior for junctions. Someone invested considerable effort in order to emulate Unix behavior for the interaction of mount points and relative symlinks.)
 
> This approach still walks backwards relying on `_getfinalpathname` - the primary difference being that it will also `readlink` to follow "real-enough" paths that can't be resolved by the IO manager. So I think this is good enough for the case where `_getfinalpathname` fails, so that we end up with 

I know. I'm saying that the `readlink` part is a can of worms that can be skipped, and I think it should be removed.

Reply:
> Thinking more about the failure case, the most likely one is a dangling relative symlink that goes up multiple levels past another symlink, which would result in the relative symlink being joined to the parent directory without resolving the parent's one first.

It could be any type of reparse point, not just a symlink. And the reparse point could be in the parent path of the relative link or actually in the target of the symlink. 

> Given how hard that scenario would be to construct, and the fact that it's purely theoretical and AFAIK nobody has ever hit it, I'm prepared to declare it a known limitation.

I would not. If a failure at some subsequent point cause `_getfinalpathname` to fail, we should honestly and accurately return as much of the real path that we can parse. We should not make radical changes to the final path but cutting reparse points out that should be traversed and if traversed would take path parsing down a completely different path.

Reply:
> I know. I'm saying that the readlink part is a can of worms that can be skipped, and I think it should be removed.

What if we do exactly one `_readlink_deep` call, immediately before we return from `_getfinalpathname_nonstrict` in the `except` clause? So at most it'll be used once, and only on something that can't give us a final name.

Any other change will require updating expected tests, and will break assumptions like:

```python
if lexists(path) and not exists(path):
    assert not lexists(realpath(path))
```